test/slam_manifold_preintegration_update_test.cc

```diff
@@ -0,0 +1,303 @@
+
+  /**
+   * Returns a matrix whose (count) colums are samples from the distribution.
+   */
+  template <typename Generator>
+  Eigen::MatrixXd Sample(Generator& gen, const int count) {
+    Eigen::MatrixXd rand = Eigen::MatrixXd::NullaryExpr(vars, count, [&]() { return dist(gen); });
+    return L * rand;
+  }
+};
+
+/**
+ * Calculates the covariance of the columns of samples.
+ */
+Eigen::MatrixXd Covariance(const Eigen::MatrixXd& samples) {
+  const long int sample_size = samples.cols();
+  const Eigen::VectorXd avg_col = samples.rowwise().sum() / sample_size;
+  const Eigen::MatrixXd meanless_samples = samples.colwise() - avg_col;
+  return (meanless_samples * meanless_samples.transpose()) / sample_size;
+}
+
+TEST_CASE("Test MultiVarNormalDist generates samples with correct covariance", "[slam]") {
+  std::mt19937 gen(170);
+
+  // Create covar, a symmetric positive definite matrix
+  // NOTE(brad): This matrix is positive semi-definite because the diagonal entries all
+  // have values of at least 2.0, and the max value of an off-diagonal entry is .2. Thus
+  // within a row, the sum of the non-diagonal entries is capped by 0.2 * 8 = 1.6. Thus by the
+  // Gershgorin circle theorem, any eigen value of covar is at least 0.4 (i.e., positive).
+  std::uniform_real_distribution<double> uniform_dist(0.01, 0.1);
+  Eigen::MatrixXd covar = Eigen::MatrixXd::NullaryExpr(9, 9, [&]() { return uniform_dist(gen); });
+  covar += covar.transpose().eval();
+  covar.diagonal().array() += 2.0;
+
+  MultiVarNormalDist dist(covar);
+
+  const Eigen::MatrixXd samples = dist.Sample(gen, 1 << 23);
+  const Eigen::Matrix<double, 9, 9> calculated_covar = Covariance(samples);
+
+  for (int col = 0; col < 9; col++) {
+    for (int row = 0; row < 9; row++) {
+      const double covar_rc = covar(row, col);
+      const double difference = covar_rc - calculated_covar(row, col);
+      // NOTE(brad): trade-off between sample count (time) and accuracy of calculated_covar
+      CHECK(difference * difference < 9e-3 * covar_rc * covar_rc);
+    }
+  }
+}
+
+namespace example {
+static const Eigen::Vector3d accel_bias = {3.4, 1.6, -5.9};
+static const Eigen::Vector3d gyro_bias = {1.2, -2.4, 0.5};
+static const Eigen::Matrix3d accel_cov =
+    (Eigen::Matrix3d() << 7e-5, 1e-7, 1e-7, 1e-7, 7e-5, 1e-7, 1e-7, 1e-7, 7e-5).finished();
+static const Eigen::Matrix3d gyro_cov =
+    (Eigen::Matrix3d() << 1e-3, 1e-5, 1e-5, 1e-5, 1e-3, 1e-5, 1e-5, 1e-5, 1e-3).finished();
+static const Eigen::Vector3d true_gyro = Eigen::Vector3d::Constant(1.2);
+static const Eigen::Vector3d true_accel = Eigen::Vector3d::Constant(4.3);
+}  // namespace example
+
+TEST_CASE("Test ImuPreintegator.covariance", "[slam]") {
+  // In order to test that the computed covariance is correct, we sample the values many
+  // time to numerically calculate the covariance to compare.
+  using M99 = Eigen::Matrix<double, 9, 9>;
+  // Parameters
+  const double dt = 1e-4;
+  const int measurements_per_sample = 130;
```

Comment:
Well, the amount of time spent (unsurprisingly) scales linearly with `measurements_per_sample`. The product of `measurements_per_sample` and `dt` is the time between keyframes, so as is we only get 130 ms between key frames (and I would like the test to actually be good for realistic values).

As to the aspect of the measurement count affecting how many samples are required to have a tight enough error for the test to be non-flakey (as a function of random variable), a quick poking around actually showed that a smaller measurement count increases the relative error between the sample covariance and the predicted covariance. Perhaps this is because if we have more measurements from our accel and gryo distributions for each sample?

Anyway, originally `1 << 20` samples was actually too few samples to not get a lot of failures when I changed the random seed (instead about `1 << 23` samples were needed). But after correctly using local-coordinates rather than `FromTangent` to calculate the `DR` sample, I was able to reduce the number of samples to `1 << 19` and still reliably have the test pass regardless of the random seed. So hopefully that's a large enough reduction.`

With 130 measurements per sample, and `1 << 19` samples, this test case (on my machine) takes 4.4 seconds running on 16 threads, 5.2 seconds running on 8 threads, and 38.3 seconds running on a single thread. Not sure if any of those are too long.

Reply:
I think we're going to run it single-threaded, 38.3s is ok but does seem kind long.  I think for checking the covariance is correct it isn't super important to use something we think is perfectly realistic.  I also think a 1e-3 dt is quite realistic.   How does the number of samples required scale with measurements_per_sample?  If we could cut it down a bit so the test takes more like 10s or less single-threaded that would be better (especially if it's slower in CI than on your machine)

Reply:
Okay. Btw, if I remember right, the larger the `dt` the more samples needed for the covariance to converge. But I still set `dt` to `1e-3`, as I thought it a more realistic value (I think I did `1e-4` before because I had a hard time getting the test to converge in a reasonable number of samples when I still had the incorrect sample for `DR`).

But there are essentially 3 variables that can be modified: (4 if you count dt), `measurements_per_sample`, `sample_count`, and the constant used to calculate the `weighted_relative_error`.

When the test fails, it is invariably in the off-diagonal portion of the `DR-DR` portion (which tends towards 0), with the calculated portion being closer to zero than the sampled version. So, the calculated off-diagonal portion might be `3e-10` while the sampled portion might be `1.7e-7` (in a failed test) and the largest covariance values being `4e-5` (which do agree nicely).

Increasing the constant used in the `weighted_relative_error` (adding a fraction of the max covariance value to the denominator to avoid dividing by 0 while at the same time weighting the larger values more heavily than the small values) is the easiest way to get the test passing.

Anyway, when I doubled this constant from 3% to 6%, I was able to reduce the sample count to `1 << 18`, the measurements per sample to 50, and thus the single threaded time to 7.5 seconds.

I could poke around looking for some other set of parameters, but I know that when the `measurements_per_sample` was only 40, the test wouldn't reliably pass when I changed the random seed.

Anyway, given that I've gotten it sub 10 seconds with plausibly reasonable values, I'm not sure if it's worth looking into further.

Reply:
Yeah I agree I don't really think it's worth tuning further.   The other knob here would be the actual covariances of all the input variables - in theory if we make them smaller everything should be a better approximation, the fact that you mention a systematic error like this points sort of in that direction:

> When the test fails, it is invariably in the off-diagonal portion of the DR-DR portion (which tends towards 0), with the calculated portion being closer to zero than the sampled version. So, the calculated off-diagonal portion might be 3e-10 while the sampled portion might be 1.7e-7 (in a failed test) and the largest covariance values being 4e-5 (which do agree nicely).

In particular we don't expect from the math that the predicted covariance should match the actual covariance up to numerical precision except in the limit as the covariances all go to zero, i.e. it might be a more useful unit test to use smaller covariances (even if we don't think they're realistic) where results should be exactly correct

Reply:
Oh, that's actually a good point. If test time weren't an issue, I would definitely want a test like that. The only thing that sticks out to me, though, is that I would still like a test with realistic values, because at the end of the day those are the conditions we want to ensure it works under.